cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -318,7 +318,7 @@ std::unique_ptr<cudf::column> gather(
 
   // check total size is not too large
   size_t const total_bytes = thrust::transform_reduce(
-    rmm::exec_policy(stream),
+    rmm::exec_policy_nosync(stream),
```

Comment:
We need the result from this returned so this will require a sync inside reduce.

Reply:
Thrust handles that for us. If a sync is required for the algorithm's return value (or some other part of its correctness), Thrust is responsible for the sync regardless of the execution policy. `nosync` really means "avoid syncing if possible."

Reply:
Reference: https://github.com/NVIDIA/thrust/releases/tag/1.16.0

> `par_nosync` is a hint to the Thrust execution engine that any non-essential internal synchronizations should be skipped and that an explicit synchronization will be performed by the caller before accessing results.

(The return value of a reduction is considered an essential synchronization.)

Reply:
I'm thinking a comment may help here?
I'm worried for the future generations (including myself) who see this (or forgot this).

Reply:
I think there are a couple underlying issues here and a couple of approaches to address them. First, `par_nosync` is pretty new, so not all Thrust developers have begun to use it and know its conventions. I think that once developers know the conventions and recognize that `nosync` is a safe choice in many cases in libcudf (but not all cases!), this will not be a point of confusion. I plan to make `nosync` changes across the entire libcudf codebase over time, so I am unsure if a code comment in _every_ location is appropriate. There are around 60 instances of `thrust::reduce` alone -- and quite a few other algorithms fall into the same boat of mandating a final synchronization for host value return. Instead, I would propose expanding our developer docs on stream synchronization to explain when `nosync` is (or is not) appropriate.

Reply:
Ok. I suppose looking at this code it certainly it appears that an internal synch _must_ be occurring otherwise the reduce would not return the correct result.

Reply:
> Instead, I would propose expanding our developer docs on stream synchronization to explain when nosync is (or is not) appropriate.

Maybe a better code talk as well? :) 
The explanation in the release notes is not very detailed. Does this mean that it will synchronize only when returning results on the host?

Reply:
Yup! That's a great idea. I have seen a better explanation than in those release notes someplace (perhaps the PR where `nosync` was introduced?) but I didn't find it last time I looked. I'll sign up for a future Better Code talk.

Reply:
@vuule November 30. Mark your calendar. 😉